caravel/views.py

```diff
@@ -1588,19 +1588,53 @@ def activity_per_day(self):
 
     @api
     @has_access_api
-    @expose("/tables/<db_id>/<schema>")
+    @expose("/schemas/<db_id>")
+    @cached_view(timeout=600)
+    def schemas(self, db_id):
+        # db_id = request.args.get('db_id')
+        database = (
+            db.session
+            .query(models.Database)
+            .filter_by(id=db_id)
+            .one()
+        )
+        return Response(
+            json.dumps({'schemas': database.all_schema_names()}),
+            mimetype="application/json")
+
+    @api
+    @has_access_api
+    @expose("/tables/<db_id>/<schema>/")
+    @cached_view(timeout=600)
     def tables(self, db_id, schema):
         """endpoint to power the calendar heatmap on the welcome page"""
-        schema = None if schema in ('null', 'undefined') else schema
+        schema = utils.js_string_to_python(schema)
+        substr = utils.js_string_to_python(request.args.get('substr'))
         database = (
             db.session
             .query(models.Database)
             .filter_by(id=db_id)
             .one()
         )
+        table_names = database.all_table_names(schema)
+        view_names = database.all_view_names(schema)
+        if substr:
+            table_names = [tn for tn in table_names if substr in tn]
+            view_names = [vn for vn in view_names if substr in vn]
+
+        max_items = config.get('MAX_TABLE_NAMES') or len(table_names)
+        total_items = len(table_names) + len(view_names)
+        max_tables = len(table_names)
+        max_views = len(view_names)
+        if total_items:
+            max_tables = max_items * len(table_names) // total_items
+            max_views = max_items * len(view_names) // total_items
+
         payload = {
-            'tables': database.all_table_names(schema),
-            'views': database.all_view_names(schema),
+            'tables': table_names[:max_tables],
+            'tables_length': len(table_names),
```

Comment:
there are 2 things here.
1 - prorate # of views / tables to fetch, it makes code complicated
2 - send # of tables / views to the UI to display the total number